File: gems/poller/src/client/signature.go
package client

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Proving we hold a PAT without ever sending it.
//
// The PAT used to ride in every request as `Authorization: Bearer`, so one
// intercepted request was total control of that controller. Now each request
// carries an HMAC over what the request IS — method, path, query, body — plus
// a timestamp and nonce that make it good once.
//
// The key is the PAT's sha256 as a 64-char lowercase hex STRING, hashed as its
// UTF-8 bytes rather than the decoded digest: "decode it first" is one more
// thing two languages can disagree about, and this is the same string the
// controller already stores.
//
// This file must agree BYTE FOR BYTE with app/services/voodu/signature.rb.
// Where they drift is query encoding, and it drifts into an intermittent 401
// that gets blamed on the network — so both suites pin the same vector.
const (
	sigScheme  = "Voodu"
	sigVersion = "v1"
)

// SignatureHeader builds the Authorization value for one request.
func SignatureHeader(pat, method, path string, query url.Values, body []byte) (string, error) {
	nonce, err := newNonce()
	if err != nil {
		return "", err
	}

	return signatureHeaderAt(pat, method, path, query, body, time.Now().Unix(), nonce), nil
}

// Separated so tests can pin a timestamp and nonce against the Ruby vector.
func signatureHeaderAt(pat, method, path string, query url.Values, body []byte, ts int64, nonce string) string {
	stamp := strconv.FormatInt(ts, 10)
	canonical := canonicalString(method, path, query, body, stamp, nonce)
	sig := signCanonical(patKey(pat), canonical)

	return fmt.Sprintf("%s id=%s, ts=%s, nonce=%s, sig=%s", sigScheme, patID(pat), stamp, nonce, sig)
}

func canonicalString(method, path string, query url.Values, body []byte, ts, nonce string) string {
	bodySum := sha256.Sum256(body)

	return strings.Join([]string{
		sigVersion,
		strings.ToUpper(method),
		path,
		canonicalQuery(query),
		hex.EncodeToString(bodySum[:]),
		ts,
		nonce,
	}, "\n")
}

// Sorted by name then value, each side percent-encoded with RFC 3986
// unreserved characters left alone. Specified rather than delegated to
// url.Values.Encode(), which sorts by key but leaves repeated values in
// insertion order and escapes a space as `+`.
func canonicalQuery(query url.Values) string {
	if len(query) == 0 {
		return ""
	}

	pairs := make([]string, 0, len(query))

	for key, values := range query {
		for _, value := range values {
			pairs = append(pairs, escapeRFC3986(key)+"="+escapeRFC3986(value))
		}
	}

	sort.Strings(pairs)

	return strings.Join(pairs, "&")
}

// url.QueryEscape turns a space into `+`; RFC 3986 wants `%20`. Ruby's
// ERB::Util.url_encode does the latter, so this matches it.
func escapeRFC3986(value string) string {
	escaped := url.QueryEscape(value)
	escaped = strings.ReplaceAll(escaped, "+", "%20")
	// QueryEscape leaves these alone in some Go versions and escapes them in
	// others; RFC 3986 calls all four unreserved.
	replacer := strings.NewReplacer("%7E", "~", "%2A", "*")

	return replacer.Replace(escaped)
}

func signCanonical(key, canonical string) string {
	mac := hmac.New(sha256.New, []byte(key))
	mac.Write([]byte(canonical))

	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

// The 64-char lowercase hex of sha256(plain) — the same string the controller
// holds as PAT.HashHex.
func patKey(pat string) string {
	sum := sha256.Sum256([]byte(pat))

	return hex.EncodeToString(sum[:])
}

// The token is `pat_` + 28 base64url characters; the first 6 of those are the
// public ID the controller keys its record by. Mirrors ParsePATToken in the
// controller, length and alphabet checks included — sending an id it will
// reject is a 401 with no explanation rather than a clean local failure.
const (
	tokenPrefix  = "pat_"
	tokenBodyLen = 28
	tokenIDLen   = 6
)

func patID(pat string) string {
	if !strings.HasPrefix(pat, tokenPrefix) {
		return ""
	}

	body := strings.TrimPrefix(pat, tokenPrefix)
	if len(body) != tokenBodyLen {
		return ""
	}

	for i := 0; i < len(body); i++ {
		c := body[i]
		if !(c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '-' || c == '_') {
			return ""
		}
	}

	return body[:tokenIDLen]
}

func newNonce() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", fmt.Errorf("signature nonce: %w", err)
	}

	return hex.EncodeToString(buf), nil
}

// SignRequest sets the Authorization header on an already-built request.
//
// Reads the method, path and query off the request itself rather than taking
// them as arguments: the signature has to cover what is actually sent, and a
// caller passing a path that has since been rewritten produces a 401 that
// looks like an auth problem and is a plumbing problem.
//
// Bodies on this plane are always empty (every call is a GET or a bodiless
// POST), so nil is signed. A future call with a body must hash the exact bytes
// it sends, not a re-serialisation of them.
func SignRequest(req *http.Request, pat string) error {
	header, err := signRequestAt(req, pat, time.Now().Unix())
	if err != nil {
		return err
	}

	req.Header.Set("Authorization", header)

	return nil
}

func signRequestAt(req *http.Request, pat string, ts int64) (string, error) {
	nonce, err := newNonce()
	if err != nil {
		return "", err
	}

	return signatureHeaderAt(pat, req.Method, req.URL.Path, req.URL.Query(), nil, ts, nonce), nil
}

File: app/services/voodu/signature.rb
# frozen_string_literal: true

require "openssl"
require "securerandom"

module Voodu
  # Signature — proving we hold a PAT without ever sending it.
  #
  # The PAT used to ride in every request as `Authorization: Bearer`, so a
  # single intercepted request was total control of that controller: deploy,
  # exec, logs. On a private network that is tolerable. Across the internet,
  # which is what the hosted SaaS does, it is not.
  #
  # So the secret stops travelling. Each request carries an HMAC over what the
  # request IS — method, path, query, body, and a timestamp and nonce that make
  # it good once. Capture it and you have that one request, for a few minutes,
  # and no way to turn a read into a restart.
  #
  # THE KEY COSTS NOTHING, which is why this shape was chosen. The client holds
  # the plain PAT and can hash it; the controller already stores exactly that
  # hash (PAT.HashHex). Both sides can compute the key, and no PAT has to be
  # re-issued.
  #
  # This protects the CREDENTIAL, not the content: bodies and responses still
  # travel in the clear. Only TLS fixes that.
  #
  # Verified byte-for-byte against the Go implementation in
  # gems/poller/src/client/signature.go — see the shared vector in the tests.
  # Two languages agreeing on query encoding is the part that breaks, and it
  # breaks as an intermittent 401 that gets blamed on the network.
  module Signature
    SCHEME = "Voodu"
    VERSION = "v1"

    # Somebody else's clock is not ours to trust to the second. Same leeway the
    # licence verifier uses.
    SKEW = 300

    module_function

    # sign — the header value for one request.
    def header(pat:, method:, path:, query: nil, body: nil, now: Time.now, nonce: SecureRandom.hex(16))
      id = pat_id(pat)
      ts = now.to_i.to_s
      canonical = canonical_string(method: method, path: path, query: query, body: body, ts: ts, nonce: nonce)

      "#{SCHEME} id=#{id}, ts=#{ts}, nonce=#{nonce}, sig=#{sign(key_for(pat), canonical)}"
    end

    # The key is the PAT's sha256 as a 64-char lowercase hex STRING, hashed as
    # its UTF-8 bytes rather than the decoded digest. "Decode it first" is one
    # more thing two languages can disagree about, and this one is free to
    # verify: it is the same string the controller has on disk.
    def key_for(pat) = OpenSSL::Digest::SHA256.hexdigest(pat.to_s)

    def sign(key, canonical)
      digest = OpenSSL::HMAC.digest("SHA256", key, canonical)

      Base64.urlsafe_encode64(digest, padding: false)
    end

    # The token is `pat_` + 28 base64url characters; the first 6 of those are
    # the public ID the controller keys its record by. Mirrors ParsePATToken in
    # internal/controller/pat.go — including the length and alphabet checks,
    # because sending an id the controller will reject is a 401 with no
    # explanation rather than a clean local failure.
    TOKEN_PREFIX = "pat_"
    TOKEN_BODY_LEN = 28
    TOKEN_ID_LEN = 6
    TOKEN_ALPHABET = /\A[A-Za-z0-9_-]+\z/

    def pat_id(pat)
      body = pat.to_s.delete_prefix(TOKEN_PREFIX)
      return "" unless pat.to_s.start_with?(TOKEN_PREFIX)
      return "" unless body.length == TOKEN_BODY_LEN && TOKEN_ALPHABET.match?(body)

      body[0, TOKEN_ID_LEN]
    end

    # Newline-joined, no trailing newline. Every field is fixed-position, so a
    # value containing a newline cannot shift the meaning of the next one — the
    # method, timestamp and nonce cannot contain one, and path and query are
    # percent-encoded before they get here.
    def canonical_string(method:, path:, ts:, nonce:, query: nil, body: nil)
      [
        VERSION,
        method.to_s.upcase,
        path.to_s,
        canonical_query(query),
        OpenSSL::Digest::SHA256.hexdigest(body.to_s),
        ts.to_s,
        nonce.to_s
      ].join("\n")
    end

    # Sorted by name then value, each side percent-encoded with RFC 3986
    # unreserved characters left alone, joined `k=v` with `&`.
    #
    # Specified rather than delegated to whatever the HTTP library does,
    # because this is exactly where two languages drift: Ruby escapes a space
    # as `+` in one API and `%20` in another, and Go's url.Values sorts by key
    # but not by value for repeated keys.
    def canonical_query(query)
      pairs = normalise_query(query)
      return "" if pairs.empty?

      pairs
        .map { |k, v| [escape(k), escape(v)] }
        .sort
        .map { |k, v| "#{k}=#{v}" }
        .join("&")
    end

    def normalise_query(query)
      case query
      when nil, "" then []
      when Hash then query.flat_map { |k, v| Array(v).map { |one| [k.to_s, one.to_s] } }
      when String then URI.decode_www_form(query).map { |k, v| [k.to_s, v.to_s] }
      else []
      end
    end

    # CGI.escape turns a space into `+`; RFC 3986 wants `%20`. Everything the
    # PAT plane sends is ASCII, but relying on that is how the first non-ASCII
    # pod name becomes a 401.
    def escape(value) = ERB::Util.url_encode(value.to_s)
  end
end
